src/hir/ambig-unambig-ty-and-consts.md

```diff
@@ -0,0 +1,63 @@
+# Ambig/Unambig Types and Consts
+
+Types and Consts args in the HIR can be in two kinds of positions "ambig" or "unambig". Ambig positions are where
+it would be valid to parse either a type or a const, unambig positions are where only one kind would be valid to
+parse.
+
+```rust
+fn func<T, const N: usize,>(arg: T) {
+    //                           ^ Unambig type position
+    let a: _ = arg; 
+    //     ^ Unambig type position
+
+    func::<T, N>(arg);
+    //     ^  ^
+    //     ^^^^ Ambig position 
+
+    let _: [u8; 10];
+    //      ^^  ^^ Unambig const position
+    //      ^^ Unambig type position
+}
+
+```
+
+Most types/consts in ambig positions are able to be disambiguated as either a type or const during either parsing or ast-lowering.
+Currently the only exception to this is inferred generic arguments in path segments. In `Foo<_>` it is not clear whether the `_` argument is an
+inferred type argument, or an inferred const argument.
+
+In unambig positions, inferred arguments are represented with [`hir::TyKind::Infer`][ty_infer] or [`hir::ConstArgKind::Infer`][const_infer] depending on whether it is a type or const position respectively.
+In ambig positions, inferred arguments are represented with `hir::GenericArg::Infer`.
+
+A naive implementation of this structure would result in there being potentially 5 places where an inferred type/const could be found in the HIR if you just looked at the types:
+- In unambig type position as a `hir::TyKind::Infer`
+- In unambig const arg position as a `hir::ConstArgKind::Infer`
+- In an ambig position as a [`GenericArg::Type(TyKind::Infer)`][generic_arg_ty]
+- In an ambig position as a [`GenericArg::Const(ConstArgKind::Infer)`][generic_arg_const]
+- In an ambig position as a [`GenericArg::Infer`][generic_arg_infer]
```

Comment:
Maybe clarify here that points 3 and 4 are the cases where there was enough other information around to disambiguate into either a type or a const?